coderd/httpmw/provisionerdaemon.go

```diff
@@ -0,0 +1,86 @@
+package httpmw
+
+import (
+	"context"
+	"crypto/subtle"
+	"net/http"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+	"github.com/coder/coder/v2/coderd/httpapi"
+	"github.com/coder/coder/v2/codersdk"
+)
+
+type provisionerDaemonContextKey struct{}
+
+func ProvisionerDaemonAuthenticated(r *http.Request) bool {
+	proxy, ok := r.Context().Value(provisionerDaemonContextKey{}).(bool)
+	return ok && proxy
+}
+
+type ExtractProvisionerAuthConfig struct {
+	DB       database.Store
+	Optional bool
+}
+
+func ExtractProvisionerDaemonAuthenticated(opts ExtractProvisionerAuthConfig, psk string) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
+
+			handleOptional := func(code int, response codersdk.Response) {
+				if opts.Optional {
+					next.ServeHTTP(w, r)
+					return
+				}
+				httpapi.Write(ctx, w, code, response)
+			}
+
+			if psk == "" {
+				// No psk means external provisioner daemons are not allowed.
+				// So their auth is not valid.
+				handleOptional(http.StatusBadRequest, codersdk.Response{
+					Message: "External provisioner daemons not enabled",
+				})
+				return
+			}
+
+			token := r.Header.Get(codersdk.ProvisionerDaemonPSK)
+			if token == "" {
+				handleOptional(http.StatusUnauthorized, codersdk.Response{
+					Message: "provisioner daemon auth token required",
+				})
+				return
+			}
+
+			if subtle.ConstantTimeCompare([]byte(token), []byte(psk)) != 1 {
+				handleOptional(http.StatusUnauthorized, codersdk.Response{
+					Message: "provisioner daemon auth token invalid",
+				})
+				return
+			}
+
+			// The PSK does not indicate a specific provisioner daemon. So just
+			// store a boolean so the caller can check if the request is from an
+			// authenticated provisioner daemon.
+			ctx = context.WithValue(ctx, provisionerDaemonContextKey{}, true)
+			// nolint:gocritic // Authenticating as a provisioner daemon.
+			ctx = dbauthz.AsProvisionerd(ctx)
+			subj, ok := dbauthz.ActorFromContext(ctx)
+			if !ok {
+				// This should never happen
+				httpapi.InternalServerError(w, xerrors.New("developer error: ExtractProvisionerDaemonAuth missing rbac actor"))
+			}
+
+			// Use the same subject for the userAuthKey
+			ctx = context.WithValue(ctx, userAuthKey{}, Authorization{
```

Comment:
1. why do we need this for provisioner daemons?

2. more generally, why do we store the `rbac.Subject` under 2 different context keys (`authContextKey{}` and `userAuthKey{}`)?

3. what is the difference between an "actor" and a "subject"? Are they interchangeable ideas and we're just inconsistent in how we refer to them in the code?

Reply:
These should be merged into 1 context value. 

It exists because `dbauthz` needed a context, and it did not feel right to use the `httpmw` functions for it. So it created it's own.


We should standardize on the `dbauthz` context, or place one in like `/rbac`.

------

1. I was just being consistent with the other auths. If I omit this, then the `HTTPAuthorizer` will not work: https://github.com/coder/coder/blob/0656d80f07245df51d538ac1cb51ef9c4828c18f/coderd/authorize.go#L66-L66. This is used in the provisioner handlers.
2. Legacy reasons when `dbauthz` was just an experiment, and not always enabled.
3. Interchangeable. One affects http handlers, one affects dbauthz.

I just made an issue tracking it here: https://github.com/coder/coder/issues/12363